### ISSUE TITLE
[BUG] Agent crashed again and again when ingress backend is not a service, a resource for example.

### ISSUE THREAD

**Issue by jaggerwang** [rancher/rancher#41736]
**Rancher Server Setup**
- Rancher version: v2.6.3
- Installation option (Docker install/Helm Chart): Helm Chart
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): AWS EKS 1.21
- Proxy/Cert Details:

**Information about the Cluster**
- Kubernetes version: Aliyun ACK 1.22.15-aliyun.1
- Cluster Type (Local/Downstream):
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider): Hosted

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
  - If custom, define the set of permissions: Admin

**Describe the bug**
After install the open source gateway [Higress](https://github.com/alibaba/higress) in the cluster, rancher agent crashed again and again. The supporter of Higress says that this is because rancher agent not consider  the condition when ingress backend is not a service, such as a resource, and they already submit a [PR](https://github.com/rancher/rancher/pull/41726) to fix this bug on v2.7, but we are using v2.6.3, can you guys fix it both on v2.7 and v2.6 as quickly as possible. 

Here is the logs of agent:
```
time="2023-06-01T09:08:24Z" level=info msg="Watching metadata for management.cattle.io/v3, Kind=Token"
time="2023-06-01T09:08:24Z" level=info msg="Watching metadata for flowcontrol.apiserver.k8s.io/v1beta1, Kind=PriorityLevelConfiguration"
time="2023-06-01T09:08:24Z" level=info msg="Watching metadata for storage.k8s.io/v1beta1, Kind=CSIStorageCapacity"
time="2023-06-01T09:08:24Z" level=info msg="Watching metadata for bookkeeper.streamnative.io/v1alpha1, Kind=BookKeeperCluster"
time="2023-06-01T09:08:24Z" level=info msg="Watching metadata for management.cattle.io/v3, Kind=AuthConfig"
time="2023-06-01T09:08:24Z" level=info msg="Watching metadata for management.cattle.io/v3, Kind=Group"
time="2023-06-01T09:08:24Z" level=info msg="Watching metadata for rbac.authorization.k8s.io/v1, Kind=ClusterRole"
time="2023-06-01T09:08:24Z" level=info msg="Watching metadata for pingcap.com/v1alpha1, Kind=DMCluster"
time="2023-06-01T09:08:24Z" level=info msg="Watching metadata for networking.k8s.io/v1, Kind=IngressClass"
W0601 09:08:24.894289      39 warnings.go:80] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0601 09:08:25.270026      39 warnings.go:80] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
I0601 09:09:11.278977      39 leaderelection.go:258] successfully acquired lease kube-system/cattle-controllers
time="2023-06-01T09:09:12Z" level=info msg="Steve auth startup complete"
time="2023-06-01T09:09:12Z" level=info msg="Registering namespaceHandler for adding labels "
W0601 09:09:13.204380      39 warnings.go:80] batch/v1beta1 CronJob is deprecated in v1.21+, unavailable in v1.25+; use batch/v1 CronJob
W0601 09:09:13.280049      39 warnings.go:80] batch/v1beta1 CronJob is deprecated in v1.21+, unavailable in v1.25+; use batch/v1 CronJob
time="2023-06-01T09:09:13Z" level=info msg="Starting catalog.cattle.io/v1, Kind=Operation controller"
time="2023-06-01T09:09:13Z" level=info msg="Starting admissionregistration.k8s.io/v1, Kind=ValidatingWebhookConfiguration controller"
time="2023-06-01T09:09:13Z" level=info msg="Starting apps/v1, Kind=Deployment controller"
E0601 09:09:13.447021      39 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 12171 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x3ccd600, 0x6eb69c0)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.3/pkg/util/runtime/runtime.go:74 +0x95
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.3/pkg/util/runtime/runtime.go:48 +0x86
panic(0x3ccd600, 0x6eb69c0)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/rancher/rancher/pkg/controllers/managementagent/endpoints.convertIngressToServicePublicEndpointsMap(0x4d318f0, 0xc016458300, 0x0, 0xc000074026, 0x5, 0x1)
	/go/src/github.com/rancher/rancher/pkg/controllers/managementagent/endpoints/endpoints.go:312 +0x6cb
github.com/rancher/rancher/pkg/controllers/managementagent/endpoints.(*WorkloadEndpointsController).UpdateEndpoints(0xc0104cbe40, 0xc00c8d16e0, 0x1e, 0xc0164761e0, 0x3a919a0, 0xc00fd836e0)
	/go/src/github.com/rancher/rancher/pkg/controllers/managementagent/endpoints/workload_endpoints.go:89 +0x674
github.com/rancher/rancher/pkg/controllers/managementagent/workload.(*CommonController).syncDeployments(0xc000156d20, 0xc00c8d16e0, 0x1e, 0xc013151af8, 0x10, 0x7fc9d91c7108, 0x10, 0xc00fd836b0)
	/go/src/github.com/rancher/rancher/pkg/controllers/managementagent/workload/workload_common.go:135 +0x1a4
github.com/rancher/rancher/pkg/generated/norman/apps/v1.(*deploymentController).AddHandler.func1(0xc00c8d16e0, 0x1e, 0x448c440, 0xc013151af8, 0xc01226bb58, 0x3, 0x3, 0x0)
	/go/src/github.com/rancher/rancher/pkg/generated/norman/apps/v1/zz_generated_deployment_controller.go:156 +0x67
github.com/rancher/norman/controller.(*genericController).AddHandler.func1(0xc00c8d16e0, 0x1e, 0x4c80c98, 0xc013151af8, 0xc013151af8, 0x4d2e9d8, 0xc013151af8, 0x1)
	/go/pkg/mod/github.com/rancher/norman@v0.0.0-20211201154850-abe17976423e/controller/generic_controller.go:62 +0x224
github.com/rancher/lasso/pkg/controller.SharedControllerHandlerFunc.OnChange(0xc0104dbc80, 0xc00c8d16e0, 0x1e, 0x4c80c98, 0xc013151af8, 0x0, 0xc013151af8, 0x0, 0x0)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20211217013041-3c6118a30611/pkg/controller/sharedcontroller.go:29 +0x4e
github.com/rancher/lasso/pkg/controller.(*SharedHandler).OnChange(0xc00b044200, 0xc00c8d16e0, 0x1e, 0x4c80c98, 0xc013151af8, 0xc01623eb01, 0x0)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20211217013041-3c6118a30611/pkg/controller/sharedhandler.go:69 +0x14c
github.com/rancher/lasso/pkg/controller.(*controller).syncHandler(0xc0008a96b0, 0xc00c8d16e0, 0x1e, 0xc00e35ace8, 0x0)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20211217013041-3c6118a30611/pkg/controller/controller.go:215 +0xd1
github.com/rancher/lasso/pkg/controller.(*controller).processSingleItem(0xc0008a96b0, 0x3a919a0, 0xc00ec0fef0, 0x0, 0x0)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20211217013041-3c6118a30611/pkg/controller/controller.go:197 +0xe7
github.com/rancher/lasso/pkg/controller.(*controller).processNextWorkItem(0xc0008a96b0, 0x203001)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20211217013041-3c6118a30611/pkg/controller/controller.go:174 +0x54
github.com/rancher/lasso/pkg/controller.(*controller).runWorker(...)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20211217013041-3c6118a30611/pkg/controller/controller.go:163
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0140dc490)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.3/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0140dc490, 0x4c53380, 0xc016236570, 0x1, 0xc00e015f80)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.3/pkg/util/wait/wait.go:156 +0x9b
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0140dc490, 0x3b9aca00, 0x0, 0x1, 0xc00e015f80)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.3/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(0xc0140dc490, 0x3b9aca00, 0xc00e015f80)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.3/pkg/util/wait/wait.go:90 +0x4d
created by github.com/rancher/lasso/pkg/controller.(*controller).run
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20211217013041-3c6118a30611/pkg/controller/controller.go:134 +0x33b
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2f85a2b]

goroutine 12171 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.3/pkg/util/runtime/runtime.go:55 +0x109
panic(0x3ccd600, 0x6eb69c0)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/rancher/rancher/pkg/controllers/managementagent/endpoints.convertIngressToServicePublicEndpointsMap(0x4d318f0, 0xc016458300, 0x0, 0xc000074026, 0x5, 0x1)
	/go/src/github.com/rancher/rancher/pkg/controllers/managementagent/endpoints/endpoints.go:312 +0x6cb
github.com/rancher/rancher/pkg/controllers/managementagent/endpoints.(*WorkloadEndpointsController).UpdateEndpoints(0xc0104cbe40, 0xc00c8d16e0, 0x1e, 0xc0164761e0, 0x3a919a0, 0xc00fd836e0)
	/go/src/github.com/rancher/rancher/pkg/controllers/managementagent/endpoints/workload_endpoints.go:89 +0x674
github.com/rancher/rancher/pkg/controllers/managementagent/workload.(*CommonController).syncDeployments(0xc000156d20, 0xc00c8d16e0, 0x1e, 0xc013151af8, 0x10, 0x7fc9d91c7108, 0x10, 0xc00fd836b0)
	/go/src/github.com/rancher/rancher/pkg/controllers/managementagent/workload/workload_common.go:135 +0x1a4
github.com/rancher/rancher/pkg/generated/norman/apps/v1.(*deploymentController).AddHandler.func1(0xc00c8d16e0, 0x1e, 0x448c440, 0xc013151af8, 0xc01226bb58, 0x3, 0x3, 0x0)
	/go/src/github.com/rancher/rancher/pkg/generated/norman/apps/v1/zz_generated_deployment_controller.go:156 +0x67
github.com/rancher/norman/controller.(*genericController).AddHandler.func1(0xc00c8d16e0, 0x1e, 0x4c80c98, 0xc013151af8, 0xc013151af8, 0x4d2e9d8, 0xc013151af8, 0x1)
	/go/pkg/mod/github.com/rancher/norman@v0.0.0-20211201154850-abe17976423e/controller/generic_controller.go:62 +0x224
github.com/rancher/lasso/pkg/controller.SharedControllerHandlerFunc.OnChange(0xc0104dbc80, 0xc00c8d16e0, 0x1e, 0x4c80c98, 0xc013151af8, 0x0, 0xc013151af8, 0x0, 0x0)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20211217013041-3c6118a30611/pkg/controller/sharedcontroller.go:29 +0x4e
github.com/rancher/lasso/pkg/controller.(*SharedHandler).OnChange(0xc00b044200, 0xc00c8d16e0, 0x1e, 0x4c80c98, 0xc013151af8, 0xc01623eb01, 0x0)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20211217013041-3c6118a30611/pkg/controller/sharedhandler.go:69 +0x14c
github.com/rancher/lasso/pkg/controller.(*controller).syncHandler(0xc0008a96b0, 0xc00c8d16e0, 0x1e, 0xc00e35ace8, 0x0)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20211217013041-3c6118a30611/pkg/controller/controller.go:215 +0xd1
github.com/rancher/lasso/pkg/controller.(*controller).processSingleItem(0xc0008a96b0, 0x3a919a0, 0xc00ec0fef0, 0x0, 0x0)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20211217013041-3c6118a30611/pkg/controller/controller.go:197 +0xe7
github.com/rancher/lasso/pkg/controller.(*controller).processNextWorkItem(0xc0008a96b0, 0x203001)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20211217013041-3c6118a30611/pkg/controller/controller.go:174 +0x54
github.com/rancher/lasso/pkg/controller.(*controller).runWorker(...)
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20211217013041-3c6118a30611/pkg/controller/controller.go:163
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0140dc490)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.3/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0140dc490, 0x4c53380, 0xc016236570, 0x1, 0xc00e015f80)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.3/pkg/util/wait/wait.go:156 +0x9b
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0140dc490, 0x3b9aca00, 0x0, 0x1, 0xc00e015f80)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.3/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(0xc0140dc490, 0x3b9aca00, 0xc00e015f80)
	/go/pkg/mod/k8s.io/apimachinery@v0.22.3/pkg/util/wait/wait.go:90 +0x4d
created by github.com/rancher/lasso/pkg/controller.(*controller).run
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20211217013041-3c6118a30611/pkg/controller/controller.go:134 +0x33b
```

**To Reproduce**
Simply install Higress following it's [Quickstart](https://higress.io/en-us/docs/user/quickstart/) and create a test route.

**Comment by kinarashah** [comment]
@jaggerwang As a workaround, this could work if you create ingress via Rancher UI and set the path rules there, could you give that a try?

**Comment by jaggerwang** [comment]
@kinarashah ingress created by ourself is ok, but the ingress auto created by Higress is out of our control.

**Comment by jaggerwang** [comment]
Is there any plan to fix this bug, it seems just need to modify a few lines of code.

**Comment by snasovich** [comment]
@jaggerwang @johnlanni , thank you for reporting this issue / submitting PR.
We'll review it for potential inclusion in the next 2.7.x release, however please note we no longer backport non-security issues into 2.6.x as it's past EOM (end of maintenance) since a few months ago (March 1st 2023 - see https://www.suse.com/lifecycle/#rancher).

**Comment by snasovich** [comment]
@Josh-Diamond , this one is ready to test as the PR was merged late last week and we have new `v2.7-head` image since then.

**Comment by Josh-Diamond** [comment]
Ticket #41736 - Test Results - ✅ 
---

Reproduced w/ HA Helm on Rancher `v2.6.3`
1. Fresh install of Rancher `v2.6.3`
2. Install Higress following the [Quickstart](https://higress.io/en-us/docs/user/quickstart/) docs
3. Reproduced - `higress-gateway` pod errors out upon initial installation w/ `ImagePullBackOff`; dirty logs 

---

Verified w/ HA Helm on Rancher `v2.7-c00569d2aadacacce2731010c726a9ee6cb82ce8-head`:
1. Fresh install of Rancher `v2.7-head`
2. Install Higress following the [Quickstart](https://higress.io/en-us/docs/user/quickstart/) docs
3. Verified - Higress successfully installed; clean logs; no errors
4. Create a test route for higress
5. Verified - higress test route successfully created; as expected
